dango/dex/src/execute.rs

```diff
@@ -221,6 +298,128 @@ fn cancel_orders(ctx: MutableCtx, order_ids: BTreeSet<OrderId>) -> anyhow::Resul
         .add_subevents(events))
 }
 
+/// Provide liquidity to a pool. The liquidity is provided in the same ratio as the
+/// current pool reserves. The required funds MUST be sent with the message. Any
+/// funds sent in excess of the required amount will be returned to the sender.
+#[inline]
+fn provide_liquidity(ctx: MutableCtx, lp_denom: Denom) -> anyhow::Result<Response> {
+    // Get the funds from sent
+    let funds: CoinPair = ctx.funds.try_into()?;
+
+    let mut pool = POOLS.load(ctx.storage, &lp_denom)?;
+
+    // Ensure the funds are valid. They must only contain the base and quote denoms and must contain both
+    ensure!(
+        funds.has(&pool.base_denom) || funds.has(&pool.quote_denom),
+        "Invalid funds. Must send at least one coin in the pair. Sent: {:?}, {:?}, {:?}",
+        funds,
+        pool.base_denom,
+        pool.quote_denom
+    );
+
+    // Ensure the pair is registered
+    ensure!(
+        PAIRS.has(ctx.storage, (&pool.base_denom, &pool.quote_denom)),
+        "Pair not found."
+    );
+
+    // Ensure the pool has reserves
+    ensure!(
+        pool.reserves.first().amount.is_non_zero() && pool.reserves.second().amount.is_non_zero(),
+        "Cannot add liquidity to pool with zero reserves"
+    );
+
+    // Query the LP token supply
+    let lp_supply = ctx.querier.query_supply(lp_denom.clone())?;
+    assert!(lp_supply.is_non_zero(), "LP token supply is zero");
+
+    // Calculate the funds to provide and the amount of LP tokens to mint
+    let mint_ratio = pool.add_liquidity(funds.clone())?;
+    let lp_mint_amount = lp_supply.checked_mul_dec_floor(mint_ratio)?;
+
+    // Apply swap fee to unbalanced provision
+    fn abs_diff(a: Uint128, b: Uint128) -> Uint128 {
+        if a > b {
+            a - b
+        } else {
+            b - a
+        }
+    }
```

Comment:
@larry0x should we implement this on `Int` instead? or as part of the `Number` trait?